Sources/Prometheus/Prometheus.swift

```diff
@@ -45,13 +64,17 @@ public class PrometheusClient {
         named name: String,
         helpText: String? = nil,
         initialValue: T = 0,
-        withLabelType labelType: U.Type) -> Counter<T, U>
+        withLabelType labelType: U.Type) -> PromCounter<T, U>
     {
-        let counter = Counter<T, U>(name, helpText, initialValue, self)
-        prometheusQueue.async(flags: .barrier) {
+        return self.lock.withLock {
+            if let type = metricTypeMap[name] {
+                precondition(type == .counter, "Label \(name) was associated with \(type) before. Can not be used for a counter now.")
+            }
+            let counter = PromCounter<T, U>(name, helpText, initialValue, self)
+            self.metricTypeMap[name] = .counter
```

Comment:
this is good, and note that even though we want to remove the `_type` on the metric type, we do keep it here and the map shall remain as well -- this is great as it powers the protecting of reusing labels 👍